discovery/manager.go

```diff
@@ -296,6 +296,49 @@ func (m *Manager) startProvider(ctx context.Context, p *Provider) {
 
 // cleaner cleans resources associated with provider.
 func (m *Manager) cleaner(p *Provider) {
+
+	// Before deleting the targets of p from m.targets definitely, we should ask
+	// the consumers to forget about them
+
+	// this logic should go somewhere else as this will block the reload
+	// maybe add a m.droppedTargets and make sender() expire them?
+	// maybe make allGroups to that? as it keeps empty targets
+
+	tSets := map[string][]*targetgroup.Group{}
+	m.targetsMtx.Lock()
+	p.mu.RLock()
+	for pkey, tsets := range m.targets {
+		// TODO: m.subs can at most contain one sub, don't see wht the plural and why a map? (maybe easy to use?)
+		for s := range p.subs {
+			key := poolKey{s, p.name}
+			for _, tg := range tsets {
+				// The targets were discovered by the provider
+				if pkey == key {
+					// Expire them
+					// works without the Source?? What is Source used for?
```

Comment:
Am I missing sth?

Reply:
A service discovery can have multiple target groups, identified by the source.

Reply:
I thought scrape manager and Co would rely on it to only drop targets from a specific source when they receive an empty group from that source, but apparently we need to send all the targets groups every time.

So, IIUC, the source is only used for `providers->disc manager` updates.

In this case, instead of sending an empty group with only the source set to scrape manager, we can just ignore it and only send the groups with targets. And in case there are no other targets, sending an empty list would be sufficient. (I'll do more tests)

Same here:

https://github.com/prometheus/prometheus/blob/1b84c01b7626cec7f10df5ca17a40da267364f15/discovery/manager.go#L399-L404

Instead of sending empty groups, we can just ignore them, but make sure to send an empty list if there are no other targets

Reply:
consumers (scrape manager e.g.) don't seem to care about the source. An empty list is sufficient.